projects/simple/src/Example/Project.hs

```diff
@@ -1,19 +1,49 @@
-module Example.Project (topEntity, plus) where
+-- @createDomain@ below generates a warning about orphan instances, but we like
+-- our code to be warning-free.
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module Example.Project where
 
 import Clash.Prelude
 
--- | Add two numbers. Example:
+-- Create a domain with the frequency of your input clock. For this example we used
+-- 50 MHz.
+createDomain vSystem{vName="MyDomain", vPeriod=hzToPeriod 50e6}
+
+-- | 'topEntity' is Clash's equivalent of 'main' in other programming languages.
+-- Clash will look for it when compiling 'Example.Project' and translate it to
+-- HDL. While polymorphism can be used freely in Clash projects, a 'topEntity'
+-- must be monomorphic and must use non- recursive types. Or, to put it
+-- hand-wavily, a 'topEntity' must be translatable to a static number of wires.
 --
--- >>> plus 3 5
--- 8
-plus :: Signed 8 -> Signed 8 -> Signed 8
-plus a b = a + b
+-- The inputs need to be declared as parameters in order to be able to give them
+-- custom names. In other words, @topEntity = exposeClockResetEnable acc@ would
+-- prevent naming the inputs.
+topEntity ::
+  Clock MyDomain ->
+  Reset MyDomain ->
+  Enable MyDomain ->
+  Signal MyDomain (Unsigned 8) ->
+  Signal MyDomain (Unsigned 8)
+topEntity clk rst en inp = (exposeClockResetEnable accum) clk rst en inp
+-- Give names to the top entity parts
+{-# ANN topEntity
+  (Synthesize
+    { t_name = "accum"
+    , t_inputs = [ PortName "CLK"
+                 , PortName "RST"
+                 , PortName "EN"
+                 , PortName "DIN"
+                 ]
+    , t_output = PortName "DOUT"
+    }) #-}
 
--- | 'topEntity' is Clash's equivalent of 'main' in other programming
--- languages. Clash will look for it when compiling 'Example.Project'
--- and translate it to HDL. While polymorphism can be used freely in
--- Clash projects, a 'topEntity' must be monomorphic and must use non-
--- recursive types. Or, to put it hand-wavily, a 'topEntity' must be
--- translatable to a static number of wires.
-topEntity :: Signed 8 -> Signed 8 -> Signed 8
-topEntity = plus
+-- | A simple accumulator: it accumulates the inputs, resets to 0, and outputs
+-- the current state
+accum ::
+  HiddenClockResetEnable dom =>
```

Comment:
Why did you make it polymorphic?

Apart from simplicity, another argument for using a concrete domain is no pain when you a use function that `HasSynchronousReset` or `HasDefinedInitialValues`.

Reply:
Personally I would strongly recommend to only use monomorphic domains for top level functions or components that require a specific domain.

Reply:
As discussed, agreed.